R/utilities-simulation.R

```diff
@@ -140,6 +140,67 @@ runSimulation <- function(simulation, population = NULL, agingData = NULL, simul
   SimulationResults$new(results, simulation)
 }
 
+#' @title  Runs a set of simulations.
+#' @details Runs a set of simulations (only individual simulations) and returns
+#' a named list of \code{SimulationResults}. The names of the entries are the IDs of the
+#' corresponding simulation (i.e. \code{simulation$id}).
+#'
+#' @param simulations A list of \code{Simulation} objects to simulate.
+#' @param simulationRunOptions Optional instance of a \code{SimulationRunOptions} used during the simulation run.
+#' @param silentMode If \code{TRUE}, no warnings are displayed if a simulation fails.
+#' Default is \code{FALSE}.
+#'
+#' @return A list of \code{SimulationResults} objects with names being the IDs of the simulations. If a simulation fails, the result for this simulation is \code{NULL}
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' sim2 <- loadSimulation(simPath)
+#' sim3 <- loadSimulation(simPath)
+#' results <- runSimulationsConcurrently(list(sim, sim2, sim3))
+#' @export
+runSimulationsConcurrently <- function(simulations, simulationRunOptions = NULL, silentMode = FALSE) {
+  validateIsOfType(simulations, Simulation)
+  simulationRunner <- getNetTask("ConcurrentSimulationRunner")
+  if (!is.null(simulationRunOptions)) {
+    validateIsOfType(simulationRunOptions, SimulationRunOptions)
+    rClr::clrSet(simulationRunner, "SimulationRunOptions", simulationRunOptions$ref)
+  }
+
+  simulations <- c(simulations)
+  # Create an Id <-> simulation map to get the correct simulation for the results.
+  simulationsIdMap <- list()
+
+  # Create SimulationRunnerConcurrentOptions and add all simulations
+  for (simulation in simulations) {
+    simulationsIdMap[[simulation$id]] <- simulation
+    rClr::clrCall(simulationRunner, "AddSimulation", simulation$ref)
+  }
+  # Run all simulations
+  results <- rClr::clrCall(simulationRunner, "RunConcurrently")
+  # Pre-allocate lists for SimulationResult
+  simulationResults <- vector("list", length(simulations))
+  # Set the order of IDs so the results appear in the same order as simulations were provided
+  names(simulationResults) <- names(simulationsIdMap)
+
+  for (i in seq_along(results)) {
+    resultObject <- results[[i]]
+    id <- rClr::clrGet(resultObject, "Id")
+    succeeded <- rClr::clrGet(resultObject, "Succeeded")
+    if (succeeded) {
+      # Get the correct simulation and create a SimulationResults object
+      simulationResults[[id]] <- SimulationResults$new(ref = rClr::clrGet(resultObject, "Result"), simulation = simulationsIdMap[[id]])
+      next()
+    }
+    # If the simulation run failed, show a warning

```

Comment:
Should not we set simulationResults[[id]] to NULL or sthg here ? to show that something is off?

Reply:
All values are NULL after initialization of `simulationResults` and are overwritten if the result exists.